File: Chapter 02/Python/2.3.4 Performance/dashboard_after.py
from database_service import DatabaseService
from cache_service import CacheService

CACHE_TTL_SECONDS = 600  # 10 minutes

class DashboardAfter:
    def __init__(self):
        self.database_service = DatabaseService()
        self.cache = CacheService()

    def get_dashboard_summary(self, user_id):
        cache_key = f"dashboard:{user_id}"
        
        # 1. Check the FAST cache first
        cached_dashboard = self.cache.get(cache_key)
        
        if cached_dashboard:
            return cached_dashboard
            
        # 2. Cache MISS. Do the slow work...
        profile = self.database_service.get_profile(user_id)
        orders = self.database_service.get_orders(user_id)
        activity = self.database_service.get_activity(user_id)
        
        dashboard_data = {
            "profile": profile,
            "orders": orders,
            "activity": activity
        }
        
        # 3. Save the result using our constant
        self.cache.set(cache_key, dashboard_data, CACHE_TTL_SECONDS)
        
        return dashboard_data

File: Chapter 02/Java/src/main/java/com/grokkingsoftwarearchitecture/chapter02/section_2_3_4_performance/DashboardAfter.java
package com.grokkingsoftwarearchitecture.chapter02.performance;
import java.util.HashMap;
import java.util.Map;

public class DashboardAfter {
    private static final int CACHE_TTL_SECONDS = 600; // 10 minutes
    
    private final DatabaseService databaseService = new DatabaseService();
    private final CacheService cache = new CacheService();

    public Map<String, String> getDashboardSummary(String userId) {
        String cacheKey = "dashboard:" + userId;

        // 1. Check the FAST cache first
        @SuppressWarnings("unchecked")
        Map<String, String> cachedDashboard = (Map<String, String>) cache.get(cacheKey);
        
        if (cachedDashboard != null) {
            return cachedDashboard;
        }

        // 2. Cache MISS. Do the slow work...
        String profile = databaseService.getProfile(userId);
        String orders = databaseService.getOrders(userId);
        String activity = databaseService.getActivity(userId);

        Map<String, String> dashboardData = new HashMap<>();
        dashboardData.put("profile", profile);
        dashboardData.put("orders", orders);
        dashboardData.put("activity", activity);

        // 3. Save the result using our constant
        cache.set(cacheKey, dashboardData, CACHE_TTL_SECONDS);

        return dashboardData;
    }
}